pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -151,22 +151,90 @@ type DNSProxy struct {
 	rejectReply int32
 }
 
+type matchNameMatcher struct {
+	mapping map[string]struct{}
+	key     *string
+}
+
+// patternMatcherEntry is a lookup entry used to cache a compiled regex
+// and how many policies it is being used by
+type patternMatcherEntry struct {
+	matcher        *regexp.Regexp
+	referenceCount int
+}
+
+// nameMatcherEntry is a lookup entry used to cache a fqdn mapping
+// and how many policies it is being used by
+type nameMatcherEntry struct {
+	matcher        *matchNameMatcher
+	referenceCount int
+}
+
 // perEPAllow maps EndpointIDs to ports + selectors + rules
-type perEPAllow map[uint64]portToSelectorAllow
+// It also contains indexes of all the regexps and the unique
+// fqdn maps in used, as well as how many rules refer to them.
+type perEPAllow struct {
+	allowMap map[uint64]portToSelectorAllow
+
+	// mapping from regexp pattern to regexp and reference count
+	patternMatchersByPattern map[string]*patternMatcherEntry
```

Comment:
oof, naming is hard.. This should probably be `patternMatcherByPattern` to follow `nameMatcherBySignature`, or...?